ffsync/ordered_id.go

```diff
@@ -0,0 +1,85 @@
+package ffsync
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/featureform/fferr"
+)
+
+type OrderedId interface {
+	Equals(other OrderedId) bool
+	Less(other OrderedId) bool
+	String() string
+	Value() interface{} // Value returns the underlying value of the ordered ID
+	MarshalJSON() ([]byte, error)
+	UnmarshalJSON(data []byte) error
+}
+
+type Uint64OrderedId uint64
+
+func (id *Uint64OrderedId) Equals(other OrderedId) bool {
+	return id.Value() == other.Value()
+}
+
+func (id *Uint64OrderedId) Less(other OrderedId) bool {
+	if otherID, ok := other.(*Uint64OrderedId); ok {
+		return *id < *otherID
+	}
+	return false
+}
+
+func (id *Uint64OrderedId) String() string {
+	return fmt.Sprint(id.Value())
+}
+
+func (id *Uint64OrderedId) Value() interface{} {
+	return uint64(*id)
+}
+
+func (id *Uint64OrderedId) UnmarshalJSON(data []byte) error {
+	var tmp uint64
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return fferr.NewInternalError(fmt.Errorf("failed to unmarshal uint64OrderedId: %v", err))
+	}
+
+	*id = Uint64OrderedId(tmp)
+	return nil
+}
+
+func (id *Uint64OrderedId) MarshalJSON() ([]byte, error) {
+	return json.Marshal(uint64(*id))
+}
+
+type OrderedIdGenerator interface {
+	NextId(namespace string) (OrderedId, fferr.GRPCError)
+}
+
+func NewMemoryOrderedIdGenerator() OrderedIdGenerator {
+	return &memoryIdGenerator{
+		counters: make(map[string]*Uint64OrderedId),
+		mu:       sync.Mutex{},
+	}
+}
+
+type memoryIdGenerator struct {
+	counters map[string]*Uint64OrderedId
+	mu       sync.Mutex
+}
+
+func (m *memoryIdGenerator) NextId(namespace string) (OrderedId, fferr.GRPCError) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
```

Comment:
If you use a sync map you may not need this lock at all, but probably not worth it tbh